java/src/main/native/src/row_conversion.cu

```diff
@@ -868,6 +878,64 @@ copy_validity_to_rows(const size_type num_rows, const size_type num_columns,
 #endif // ASYNC_MEMCPY_SUPPORTED
 }
 
+/**
+ * @brief kernel to copy string data to JCUDF row format
+ *
+ * @tparam RowOffsetIter iterator for row offsets into the destination data
+ * @param num_rows number of rows in this portion of the table
+ * @param num_variable_columns number of columns of variable-width data
+ * @param variable_input_data variable width data column pointers
+ * @param variable_col_output_offsets output offset information for variable-width columns
+ * @param variable_col_offsets input offset information for variable-width columns
+ * @param fixed_width_row_size offset to variable-width data in a row
+ * @param row_offsets offsets for each row in output data
+ * @param batch_row_offset row start for this batch
+ * @param output_data pointer to output data for this batch
+ *
+ */
+template <typename RowOffsetIter>
+__global__ void copy_strings_to_rows(size_type const num_rows, size_type const num_variable_columns,
+                                     int8_t const **variable_input_data,
+                                     size_type const *variable_col_output_offsets,
+                                     size_type const **variable_col_offsets,
+                                     size_type fixed_width_row_size, RowOffsetIter row_offsets,
+                                     size_type const batch_row_offset, int8_t *output_data) {
+  // Each block will take a group of rows controlled by NUM_STRING_ROWS_PER_BLOCK.
+  // Each warp will copy a row at a time. The base thread will first go through column data and
+  // fill out offset/length information for the column. Then all threads of the warp will
+  // participate in the memcpy of the string data.
+  auto my_block = cooperative_groups::this_thread_block();
+  auto my_tile = cooperative_groups::tiled_partition<32>(my_block);
+#ifdef ASYNC_MEMCPY_SUPPORTED
+  cuda::barrier<cuda::thread_scope_block> block_barrier;
+#endif
+
+  auto const start_row =
+      blockIdx.x * NUM_STRING_ROWS_PER_BLOCK + my_tile.meta_group_rank() + batch_row_offset;
+  auto const end_row =
+      std::min(num_rows, static_cast<size_type>(start_row + NUM_STRING_ROWS_PER_BLOCK));
+
+  for (int row = start_row; row < end_row; row += my_tile.meta_group_size()) {
+    auto offset = fixed_width_row_size; // initial offset to variable-width data
+    auto const base_row_offset = row_offsets(row, 0);
+    for (int col = 0; col < num_variable_columns; ++col) {
+      auto const string_start_offset = variable_col_offsets[col][row];
+      auto const string_length = variable_col_offsets[col][row + 1] - string_start_offset;
+      if (my_tile.thread_rank() == 0) {
+        // write the offset/length to column
+        uint32_t *output_dest = reinterpret_cast<uint32_t *>(
+            &output_data[base_row_offset + variable_col_output_offsets[col]]);
+        *output_dest++ = offset;
+        *output_dest++ = string_length;
```

Comment:
I typically strongly object to use such statements. `*var++` is very ambiguous and error-prone: I'm not clear which operator will execute first. Please add parenthesis to clarify it, such as `(*var)++` or `*(var++)`.

Reply:
I'm the same, but with arithmetic expressions.  The more parentheses, the better.

Reply:
I made a change here, see what you think. I went with:
```
output_dest[0] = offset;
output_dest[1] = string_length;
```

Reply:
> I typically strongly object to using such statements.

You might not like the reference/example implementation Bjarne Stroustrup gave for `strcpy()` (or was it `std::copy()`?) in "The C++ Programming Language". :]